<a href="https://colab.research.google.com/github/sultanadeel/timeseries-data-classification/blob/main/M1_Assignment_Stock_Price_Prediction_with_LLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [1]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Display summary statistics
print(data.describe())


/tmp/ipython-input-1433581267.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.488414  129.035076  125.828629  127.410552  1.120920e+08
std      30.386975   30.664097   30.105382   30.415044  5.602586e+07
min      54.316944   55.316762   51.470008   55.215087  3.519590e+07
25%     112.581745  114.075479  111.125716  112.761702  7.636470e+07
50%     132.860008  133.991789  130.902108  132.681725  9.493580e+07
75%     148.053474  149.219268  146.111078  147.421474  1.296327e+08
max     178.443146  179.354901  175.609755  179.050979  4.265100e+08


In [ ]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())


YF.download() has changed argument auto_adjust default to True


[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.800474  129.350921  126.136626  127.722421  1.120920e+08
std      30.461355   30.739156   30.179073   30.489492  5.602586e+07
min      54.449894   55.452160   51.595990   55.350235  3.519590e+07
25%     112.857315  114.354702  111.397715  113.037711  7.636470e+07
50%     133.185188  134.319768  131.222483  133.006499  9.493580e+07
75%     148.415882  149.584513  146.468716  147.782312  1.296327e+08
max     178.879913  179.793905  176.039592  179.489238  4.265100e+08


In [ ]:
# Prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())



YF.download() has changed argument auto_adjust default to True


[*********************100%***********************]  1 of 1 completed


Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.800473  129.350921  126.136626  127.722421  1.120920e+08
std      30.461355   30.739156   30.179073   30.489492  5.602586e+07
min      54.449894   55.452160   51.595990   55.350235  3.519590e+07
25%     112.857307  114.354706  111.397708  113.037705  7.636470e+07
50%     133.185196  134.319791  131.222513  133.006484  9.493580e+07
75%     148.415874  149.584494  146.468701  147.782328  1.296327e+08
max     178.879898  179.793936  176.039622  179.489269  4.265100e+08


## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

In [2]:
# Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())

/tmp/ipython-input-1773250383.py:6: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.488414  129.035076  125.828629  127.410552  1.120920e+08
std      30.386975   30.664097   30.105382   30.415044  5.602586e+07
min      54.316944   55.316762   51.470008   55.215087  3.519590e+07
25%     112.581745  114.075479  111.125716  112.761702  7.636470e+07
50%     132.860008  133.991789  130.902108  132.681725  9.493580e+07
75%     148.053474  149.219268  146.111078  147.421474  1.296327e+08
max     178.443146  179.354901  175.609755  179.050979  4.265100e+08


# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.



In [ ]:
# Load historical data for AAPL for the prediction period
import yfinance as yf

prediction_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")

# Display the first few rows of the prediction data
display(prediction_data.head())

# Task
Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics. Suggest some features that might help improve a stock price prediction model. Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

## Feature engineering

### Subtask:
Calculate moving averages and technical indicators based on the historical price data.


**Reasoning**:
Calculate the moving averages and add the volume to the DataFrame as new columns.



In [3]:
# Calculate moving averages
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Display the first few rows with the new features
display(data.head())

Price,Close,High,Low,Open,Volume,MA_5,MA_10,MA_50
Ticker,AAPL,AAPL,AAPL,AAPL,AAPL,,,
Date,,,,,,,,
2020-01-02,72.538506,72.598884,71.292296,71.545882,135480400,NaN,NaN,NaN
2020-01-03,71.833305,72.594071,71.608700,71.765682,146322800,NaN,NaN,NaN
2020-01-06,72.405670,72.444313,70.703005,70.954181,118387200,NaN,NaN,NaN
2020-01-07,72.065155,72.671348,71.845377,72.415345,108872000,NaN,NaN,NaN
2020-01-08,73.224411,73.526303,71.768086,71.768086,132079200,72.413409,NaN,NaN


## Prepare data for modeling

### Subtask:
Create the target variable (tomorrow's stock price) and select the features.


**Reasoning**:
Create the target variable, remove the last row with the missing target, and select the features and target variables for the model.



In [4]:
# Create the target variable (tomorrow's close price)
data['Target'] = data['Close'].shift(-1)

# Drop the last row as it will have a NaN target
data.dropna(inplace=True)

# Define the features and target
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50']
target = 'Target'

X = data[features]
y = data[target]

# Display the first few rows of the features and target
display(X.head())
display(y.head())

Price,Open,High,Low,Close,Volume,MA_5,MA_10,MA_50
Ticker,AAPL,AAPL,AAPL,AAPL,AAPL,,,
Date,,,,,,,,
2020-03-13,64.126276,67.764834,61.235759,67.292763,370732000,65.515371,68.412666,73.941987
2020-03-16,58.572820,62.719759,58.100752,58.635765,322423600,64.355294,67.042458,73.663932
2020-03-17,59.918810,62.363881,57.713403,61.213974,324056000,62.782697,66.159807,73.451545
2020-03-18,58.045070,60.521614,57.403539,59.715466,300233600,61.390218,64.802428,73.197741
2020-03-19,59.889765,61.209135,58.732592,59.257919,271857200,61.223177,63.637023,72.941597


,Target
Date,
2020-03-13,58.635765
2020-03-16,61.213974
2020-03-17,59.715466
2020-03-18,59.257919
2020-03-19,55.495899


## Split data

### Subtask:
Split the data into training and testing sets using a 70/30 ratio.


**Reasoning**:
Split the data into training and testing sets using a 70/30 ratio.



In [5]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Print the shapes of the resulting sets
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (494, 8)
Shape of X_test: (212, 8)
Shape of y_train: (494,)
Shape of y_test: (212,)


## Build and train regression model

### Subtask:
Choose and train a regression model (e.g., Linear Regression) on the training data.


**Reasoning**:
Train a Linear Regression model using the training data.



In [6]:
from sklearn.linear_model import LinearRegression

# Instantiate the Linear Regression model
model = LinearRegression()

# Fit the model to the training data
model.fit(X_train, y_train)

LinearRegression()

## Evaluate model

### Subtask:
Predict on the test data and report performance metrics such as Mean Squared Error (MSE) and R-squared (R²).


**Reasoning**:
Make predictions on the test set and calculate the MSE and R-squared metrics to evaluate the model's performance.



In [7]:
from sklearn.metrics import mean_squared_error, r2_score

# Make predictions on the test data
y_pred = model.predict(X_test)

# Calculate Mean Squared Error (MSE)
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error (MSE): {mse}")

# Calculate R-squared (R²) score
r2 = r2_score(y_test, y_pred)
print(f"R-squared (R²) Score: {r2}")

Mean Squared Error (MSE): 6.539000137135291
R-squared (R²) Score: 0.9918026235330162


## Summary:

### Data Analysis Key Findings

*   The dataset for Apple Inc. (AAPL) stock prices from 2020-01-01 to 2023-01-01 was loaded.
*   Moving averages for 5, 10, and 50 days were calculated and added as features.
*   The volume data was available and included as a feature.
*   A target variable representing the next day's closing price was created.
*   The data was split into training (70%, 494 samples) and testing (30%, 212 samples) sets.
*   A Linear Regression model was trained on the training data.
*   The model achieved a Mean Squared Error (MSE) of 6.539 on the test set.
*   The model achieved an R-squared (R²) score of 0.9918 on the test set, indicating a high proportion of variance explained.

### Insights or Next Steps

*   While the current model shows high performance metrics, exploring other regression algorithms like Random Forest or Gradient Boosting might yield further improvements or robustness.
*   Incorporating additional technical indicators (e.g., RSI, MACD) or external factors (e.g., news sentiment, economic indicators) could potentially enhance the predictive capability of the model.


# Task
Build a random forest model to predict the direction of Apple stock price movement using historical data from 2020-01-01 to 2023-01-01 from Yahoo Finance. Include moving averages and other relevant features, split the data into 70/30 training/testing sets, and report classification metrics and feature importance.

## Feature engineering

### Subtask:
Calculate moving averages and other relevant features from the historical price data.


**Reasoning**:
Calculate the moving averages and add them to the DataFrame as new columns.



In [8]:
# Calculate moving averages
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Display the first few rows with the new features
display(data.head())

Price,Close,High,Low,Open,Volume,MA_5,MA_10,MA_50,Target
Ticker,AAPL,AAPL,AAPL,AAPL,AAPL,,,,
Date,,,,,,,,,
2020-03-13,67.292763,67.764834,61.235759,64.126276,370732000,NaN,NaN,NaN,58.635765
2020-03-16,58.635765,62.719759,58.100752,58.572820,322423600,NaN,NaN,NaN,61.213974
2020-03-17,61.213974,62.363881,57.713403,59.918810,324056000,NaN,NaN,NaN,59.715466
2020-03-18,59.715466,60.521614,57.403539,58.045070,300233600,NaN,NaN,NaN,59.257919
2020-03-19,59.257919,61.209135,58.732592,59.889765,271857200,NaN,NaN,NaN,55.495899


## Create target variable

### Subtask:
Create a binary target variable indicating whether the stock price increased or decreased on the next day.


**Reasoning**:
Calculate the daily price change and create a binary target variable based on whether the price increased or decreased.



In [9]:
# Calculate the daily price change
data['Price_Change'] = data['Close'].shift(-1) - data['Close']

# Create the binary target variable
data['Target'] = (data['Price_Change'] > 0).astype(int)

# Display the first few rows with the new features
display(data.head())

Price,Close,High,Low,Open,Volume,MA_5,MA_10,MA_50,Target,Price_Change
Ticker,AAPL,AAPL,AAPL,AAPL,AAPL,,,,,
Date,,,,,,,,,,
2020-03-13,67.292763,67.764834,61.235759,64.126276,370732000,NaN,NaN,NaN,0,-8.656998
2020-03-16,58.635765,62.719759,58.100752,58.572820,322423600,NaN,NaN,NaN,1,2.578209
2020-03-17,61.213974,62.363881,57.713403,59.918810,324056000,NaN,NaN,NaN,0,-1.498508
2020-03-18,59.715466,60.521614,57.403539,58.045070,300233600,NaN,NaN,NaN,0,-0.457546
2020-03-19,59.257919,61.209135,58.732592,59.889765,271857200,NaN,NaN,NaN,0,-3.762020


## Prepare data for modeling

### Subtask:
Select the features and the new binary target variable and remove rows with missing values.


**Reasoning**:
Drop rows with NaN values, define features and target, and create X and y DataFrames.



In [10]:
# Drop rows with NaN values
data.dropna(inplace=True)

# Define the features and target
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50']
target = 'Target'

# Create X and y
X = data[features]
y = data[target]

# Display the first few rows of X and y
display(X.head())
display(y.head())

Price,Open,High,Low,Close,Volume,MA_5,MA_10,MA_50
Ticker,AAPL,AAPL,AAPL,AAPL,AAPL,,,
Date,,,,,,,,


,Target
Date,


**Reasoning**:
The previous step resulted in empty dataframes after dropping NaN values. This indicates an issue with how the data was loaded or processed earlier. I need to reload the data and re-apply the feature engineering and target variable creation before dropping NaNs.



In [11]:
import yfinance as yf

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Calculate moving averages
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Calculate the daily price change
data['Price_Change'] = data['Close'].shift(-1) - data['Close']

# Create the binary target variable
data['Target'] = (data['Price_Change'] > 0).astype(int)

# Drop rows with NaN values
data.dropna(inplace=True)

# Define the features and target
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50']
target = 'Target'

# Create X and y
X = data[features]
y = data[target]

# Display the first few rows of X and y
display(X.head())
display(y.head())

/tmp/ipython-input-1513411903.py:4: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed


Price,Open,High,Low,Close,Volume,MA_5,MA_10,MA_50
Ticker,AAPL,AAPL,AAPL,AAPL,AAPL,,,
Date,,,,,,,,
2020-03-13,64.126276,67.764834,61.235759,67.292763,370732000,65.515371,68.412666,73.941987
2020-03-16,58.572820,62.719759,58.100752,58.635765,322423600,64.355294,67.042458,73.663932
2020-03-17,59.918810,62.363881,57.713403,61.213974,324056000,62.782697,66.159807,73.451545
2020-03-18,58.045070,60.521614,57.403539,59.715466,300233600,61.390218,64.802428,73.197741
2020-03-19,59.889765,61.209135,58.732592,59.257919,271857200,61.223177,63.637023,72.941597


,Target
Date,
2020-03-13,0
2020-03-16,1
2020-03-17,0
2020-03-18,0
2020-03-19,0


## Split data

### Subtask:
Split the data into training and testing sets.


**Reasoning**:
Split the feature and target data into training and testing sets using a 70/30 ratio and print their shapes.



In [12]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Print the shapes of the resulting sets
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (494, 8)
Shape of X_test: (212, 8)
Shape of y_train: (494,)
Shape of y_test: (212,)


## Build and train random forest model

### Subtask:
Instantiate and train a Random Forest classifier on the training data.


**Reasoning**:
Instantiate and train a Random Forest classifier on the training data.



In [13]:
from sklearn.ensemble import RandomForestClassifier

# Instantiate the Random Forest Classifier model
model = RandomForestClassifier(random_state=42)

# Fit the model to the training data
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

## Evaluate model

### Subtask:
Predict on the test data and report classification performance metrics such as Accuracy, Confusion Matrix, and Classification Report.


**Reasoning**:
Predict on the test data and calculate the classification metrics.



In [14]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Make predictions on the test data
y_pred = model.predict(X_test)

# Calculate and print Accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")

# Calculate and print Confusion Matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print(f"Confusion Matrix:\n{conf_matrix}")

# Generate and print Classification Report
class_report = classification_report(y_test, y_pred)
print(f"Classification Report:\n{class_report}")

Accuracy: 0.5094339622641509
Confusion Matrix:
[[53 49]
 [55 55]]
Classification Report:
              precision    recall  f1-score   support

           0       0.49      0.52      0.50       102
           1       0.53      0.50      0.51       110

    accuracy                           0.51       212
   macro avg       0.51      0.51      0.51       212
weighted avg       0.51      0.51      0.51       212



## Feature importance

### Subtask:
Determine and report the importance of each feature in the Random Forest model.


**Reasoning**:
Determine feature importances from the trained Random Forest model and display them.



In [15]:
import pandas as pd

# Get feature importances from the model
feature_importances = model.feature_importances_

# Create a pandas Series with feature names and their importances
feature_importances_series = pd.Series(feature_importances, index=features)

# Sort the features by importance in descending order
sorted_feature_importances = feature_importances_series.sort_values(ascending=False)

# Print the sorted feature importances
print("Feature Importances:")
print(sorted_feature_importances)

Feature Importances:
High      0.135549
Volume    0.132330
MA_50     0.125030
MA_5      0.123389
Open      0.123295
Close     0.122211
MA_10     0.120462
Low       0.117733
dtype: float64


## Summary:

### Data Analysis Key Findings

*   The data was successfully split into training and testing sets with a 70/30 ratio, resulting in 494 samples for training and 212 samples for testing.
*   A Random Forest Classifier model was trained on the training data.
*   The model achieved an accuracy of approximately 50.94% on the test set.
*   The confusion matrix and classification report indicated that the model's performance was close to random guessing for predicting the direction of stock price movement.
*   Based on feature importance analysis, the 'High' price was the most important feature in the model, followed by 'Volume', 'MA\_50', and 'MA\_5'.

### Insights or Next Steps

*   The current model performance (around 50.94% accuracy) suggests that using only the selected features and a standard Random Forest classifier is not sufficient to reliably predict the direction of Apple stock price movement.
*   Further steps could involve exploring more advanced feature engineering techniques (e.g., technical indicators like RSI, MACD), trying different machine learning models, or incorporating external factors that might influence stock prices.


## Create target variable

### Subtask:
Create a binary target variable indicating whether the stock price increased or decreased on the next day.

**Reasoning**:
Calculate the daily price change and create a binary target variable based on whether the price increased or decreased.

In [16]:
# Calculate the daily price change
data['Price_Change'] = data['Close'].shift(-1) - data['Close']

# Create the binary target variable
data['Target'] = (data['Price_Change'] > 0).astype(int)

# Display the first few rows with the new features
display(data.head())

Price,Close,High,Low,Open,Volume,MA_5,MA_10,MA_50,Price_Change,Target
Ticker,AAPL,AAPL,AAPL,AAPL,AAPL,,,,,
Date,,,,,,,,,,
2020-03-13,67.292763,67.764834,61.235759,64.126276,370732000,65.515371,68.412666,73.941987,NaN,0
2020-03-16,58.635765,62.719759,58.100752,58.572820,322423600,64.355294,67.042458,73.663932,NaN,0
2020-03-17,61.213974,62.363881,57.713403,59.918810,324056000,62.782697,66.159807,73.451545,NaN,0
2020-03-18,59.715466,60.521614,57.403539,58.045070,300233600,61.390218,64.802428,73.197741,NaN,0
2020-03-19,59.257919,61.209135,58.732592,59.889765,271857200,61.223177,63.637023,72.941597,NaN,0


## Prepare data for modeling

### Subtask:
Select the features and the new binary target variable and remove rows with missing values.

**Reasoning**:
Drop rows with NaN values, define features and target, and create X and y DataFrames.

In [17]:
# Drop rows with NaN values
data.dropna(inplace=True)

# Define the features and target
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50']
target = 'Target'

# Create X and y
X = data[features]
y = data[target]

# Display the first few rows of X and y
display(X.head())
display(y.head())

Price,Open,High,Low,Close,Volume,MA_5,MA_10,MA_50
Ticker,AAPL,AAPL,AAPL,AAPL,AAPL,,,
Date,,,,,,,,


,Target
Date,


**Reasoning**:
The previous step resulted in empty dataframes after dropping NaN values. This indicates an issue with how the data was loaded or processed earlier. I need to reload the data and re-apply the feature engineering and target variable creation before dropping NaNs.

In [18]:
import yfinance as yf

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Calculate moving averages
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Calculate the daily price change
data['Price_Change'] = data['Close'].shift(-1) - data['Close']

# Create the binary target variable
data['Target'] = (data['Price_Change'] > 0).astype(int)

# Drop rows with NaN values
data.dropna(inplace=True)

# Define the features and target
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50']
target = 'Target'

# Create X and y
X = data[features]
y = data[target]

# Display the first few rows of X and y
display(X.head())
display(y.head())

/tmp/ipython-input-1513411903.py:4: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed


Price,Open,High,Low,Close,Volume,MA_5,MA_10,MA_50
Ticker,AAPL,AAPL,AAPL,AAPL,AAPL,,,
Date,,,,,,,,
2020-03-13,64.126276,67.764834,61.235759,67.292763,370732000,65.515371,68.412666,73.941987
2020-03-16,58.572820,62.719759,58.100752,58.635765,322423600,64.355294,67.042458,73.663932
2020-03-17,59.918810,62.363881,57.713403,61.213974,324056000,62.782697,66.159807,73.451545
2020-03-18,58.045070,60.521614,57.403539,59.715466,300233600,61.390218,64.802428,73.197741
2020-03-19,59.889765,61.209135,58.732592,59.257919,271857200,61.223177,63.637023,72.941597


,Target
Date,
2020-03-13,0
2020-03-16,1
2020-03-17,0
2020-03-18,0
2020-03-19,0


## Split data

### Subtask:
Split the data into training and testing sets.

**Reasoning**:
Split the feature and target data into training and testing sets using a 70/30 ratio and print their shapes.

In [19]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Print the shapes of the resulting sets
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (494, 8)
Shape of X_test: (212, 8)
Shape of y_train: (494,)
Shape of y_test: (212,)


## Build and train random forest model

### Subtask:
Instantiate and train a Random Forest classifier on the training data.

**Reasoning**:
Instantiate and train a Random Forest classifier on the training data.

In [20]:
from sklearn.ensemble import RandomForestClassifier

# Instantiate the Random Forest Classifier model
model = RandomForestClassifier(random_state=42)

# Fit the model to the training data
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

## Evaluate model

### Subtask:
Predict on the test data and report classification performance metrics such as Accuracy, Confusion Matrix, and Classification Report.

**Reasoning**:
Predict on the test data and calculate the classification metrics.

In [21]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Make predictions on the test data
y_pred = model.predict(X_test)

# Calculate and print Accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")

# Calculate and print Confusion Matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print(f"Confusion Matrix:\n{conf_matrix}")

# Generate and print Classification Report
class_report = classification_report(y_test, y_pred)
print(f"Classification Report:\n{class_report}")

Accuracy: 0.5094339622641509
Confusion Matrix:
[[53 49]
 [55 55]]
Classification Report:
              precision    recall  f1-score   support

           0       0.49      0.52      0.50       102
           1       0.53      0.50      0.51       110

    accuracy                           0.51       212
   macro avg       0.51      0.51      0.51       212
weighted avg       0.51      0.51      0.51       212



## Feature importance

### Subtask:
Determine and report the importance of each feature in the Random Forest model.

**Reasoning**:
Determine feature importances from the trained Random Forest model and display them.

In [22]:
import pandas as pd

# Get feature importances from the model
feature_importances = model.feature_importances_

# Create a pandas Series with feature names and their importances
feature_importances_series = pd.Series(feature_importances, index=features)

# Sort the features by importance in descending order
sorted_feature_importances = feature_importances_series.sort_values(ascending=False)

# Print the sorted feature importances
print("Feature Importances:")
print(sorted_feature_importances)

Feature Importances:
High      0.135549
Volume    0.132330
MA_50     0.125030
MA_5      0.123389
Open      0.123295
Close     0.122211
MA_10     0.120462
Low       0.117733
dtype: float64


## Summary:

### Data Analysis Key Findings

* The data was successfully split into training and testing sets with a 70/30 ratio, resulting in 494 samples for training and 212 samples for testing.
* A Random Forest Classifier model was trained on the training data.
* The model achieved an accuracy of approximately 50.94% on the test set.
* The confusion matrix and classification report indicated that the model's performance was close to random guessing for predicting the direction of stock price movement.
* Based on feature importance analysis, the 'High' price was the most important feature in the model, followed by 'Volume', 'MA\_50', and 'MA\_5'.

### Insights or Next Steps

* The current model performance (around 50.94% accuracy) suggests that using only the selected features and a standard Random Forest classifier is not sufficient to reliably predict the direction of Apple stock price movement.
* Further steps could involve exploring more advanced feature engineering techniques (e.g., technical indicators like RSI, MACD), trying different machine learning models, or incorporating external factors that might influence stock prices.